src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -411,13 +412,44 @@ internal override void Clear()
         {
             lock (_locker)
             {
-                LinkedList<ProjectRootElement> oldStrongCache = _strongCache;
-                _weakCache = new WeakValueDictionary<string, ProjectRootElement>(StringComparer.OrdinalIgnoreCase);
-                _strongCache = new LinkedList<ProjectRootElement>();
+                if (Traits.Instance.EscapeHatches.AlwaysDoImmutableFilesUpToDateCheck)
+                {
+                    LinkedList<ProjectRootElement> oldStrongCache = _strongCache;
+                    _weakCache = new WeakValueDictionary<string, ProjectRootElement>(StringComparer.OrdinalIgnoreCase);
+                    _strongCache = new LinkedList<ProjectRootElement>();
 
-                foreach (ProjectRootElement projectRootElement in oldStrongCache)
+                    foreach (ProjectRootElement projectRootElement in oldStrongCache)
+                    {
+                        RaiseProjectRootElementRemovedFromStrongCache(projectRootElement);
+                    }
+                }
+                else
                 {
-                    RaiseProjectRootElementRemovedFromStrongCache(projectRootElement);
+                    // Manually iterate through LinkedList so we can remove items during this iteration
+                    for (var listNode = _strongCache.First; listNode != null;)
+                    {
+                        var nextNode = listNode.Next;
+
+                        ProjectRootElement projectRootElement = listNode.Value;
+                        // Do not remove cache of files from immutable locations.
+                        // Those are mostly SDK project files and will be most probably needed in next builds.
+                        if (!FileClassifier.Shared.IsNonModifiable(projectRootElement.FullPath))
+                        {
+                            _weakCache.Remove(projectRootElement.FullPath);
+                            _strongCache.Remove(listNode);
+                            RaiseProjectRootElementRemovedFromStrongCache(projectRootElement);
+                        }
+
+                        listNode = nextNode;
+                    }
+
+                    // From weak list remove all which is not in strong list anymore
+                    IList<string> toBeRemovedFromWeakRefs = _weakCache.Keys.Except(_strongCache.Select(i => i.FullPath)).ToList();
```

Comment:
This walk feels expensive to me. What is causing things to be removed from the strong cache other than the for above, which is removing from both?

Reply:
The other example I'm aware of is that the strong cache has a maximum size and will remove stuff randomly if we try to go over that:
https://github.com/dotnet/msbuild/blob/a0c282787d59ca9a449ded85175f39030e5077b8/src/Build/Evaluation/ProjectRootElementCache.cs#L570-L577

Reply:
Maximum size is 200 by default, but you can set it smaller.

Reply:
This may be a silly question, but where do we actually use the strong cache?

Reply:
We don't - GC does.  strong cache is there to hold strong reference to cache item so GC will not collect it. Where there is no active/strong reference GC will collect it at next Collect.
Due to heavy allocation patterns in MSBuild it means that we have mostly only 200 XMLs cached. Which often is not enough. For example OrchardCore during build have 1500 distinct XMLs read - every project have couple `directory.proj` etc...

Reply:
@rainersigwald 
> This walk feels expensive to me.
It is not really expensive. It is called only once between -restore and build, have complexity O(N+M), where N and M are about 200.

> What is causing things to be removed from the strong cache other than the for above, which is removing from both?
`Project.UnloadProjects` calls `ProjectRootElementCache.DiscardStrongReferences()`

Reply:
> GC does

So if we don't use it, we shouldn't really care about duplicates or structure, etc., right? I'm thinking two thoughts.

One is that arrays are much more space-efficient than LinkedLists; the only problem with arrays is that they're static, so we can't reallocate them over and over again cheaply to add one more element. Proposal: rather than a simple LinkedList, we could have a LinkedList of arrays. The last would be size 32; the second-to-last would be 64; the third-to-last would be 128; etc. When we want to add something, we go to the head, check if it's full, and if not, add to that. If so, add a new node in front of it, and put it there.

Second, we're putting a lot of effort into making it a semi-legit cache. I'd propose getting lazy and stripping out all the times we currently walk the list, just resetting it when we discard strong references and at the end of builds. In this case, we also wouldn't have a maximum size for the cache.

What do you think? If you think it might help, I'd be happily to (experimentally) implement it so we can measure its performance. Either way, it isn't too relevant for this particular PR, so it can be a follow-up.

Reply:
Our `ProjecRootElementCache` is LRU, Last Recently Used, cache. 
Out solution is text book of 'theoretical' correctness, the one expected on technical interviews, you can seek internet wisdom for best theoretical solutions.
Now for practical correctness, because our cache max count is just 200, it could be solved many ways, as even O(n^3) would still be acceptable. Caching and especially its invalidations are common places of nasty concurrence bugs, and I am mostly trying to not touch it, unless the gain is significant. 
Our current LRN mechanism, weak/strong, is goof enough, actually too good with a baggage of unnecessary complexity, and I do not believe that optimizing data structures will bring any practical performance gain.
However, it is good learning opportunity but for that I would recommend to isolate our logic and use `BenchmarkDotNet` tooling. If you are interesting I can give you crash course in it - it is very useful tool.

Reply:
> Out solution is text book of 'theoretical' correctness, the one expected on technical interviews, you can seek internet wisdom for best theoretical solutions.

-ish? The proof that LRU caches are at worst 2x as expensive as the ideal cache comes with two assumptions: you're using a LinkedList, and you _access_ things in the cache; they aren't just cached to prevent GC.

I'll build OrchardCore once, since that was the example you used earlier, with some extra traces to see what gets called when, then throw the results into BenchmarkDotNet and report results when I can.

Reply:
I mocked the cache calls involved in a single (clean) OC build:
|  Method |       Mean |     Error |    StdDev |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------  |------------:|----------:|----------:|--------:|-------:|------:|----------:|
| Current |  1,571.73 us | 31.399 us | 30.838 us | 76.1719 | 1.9531 |     - | 469.74 KB |
|    Mine  |    55.32 us |  0.701 us |  0.585 us | 20.8130 | 6.8970 |     - | 128.74 KB |

I'm happy showing you more precisely what exactly I did if you want to look at it together.

As far as I'm concerned, this is a massive improvement percentage-wise but not a massive improvement total time-wise, so maybe it's worth working on, since I don't think it would take _too_ long, but up to you 🙂